### PR TITLE
[Keyboard] Add kalkulators

### PR DESCRIPTION
**Pull request by georgsnarbuts**
kalkulators_front1 (1)

## Description

Added the whole code-base needed for a functioning unit. Includes a driver for the tm1638 led driver chip and also the calculator logic. VIA support as well. 

## Types of Changes

- [x ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x ] Keyboard (addition or update)
- [x ] Keymap/layout/userspace (addition or update)
- [x ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [x ] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x ] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [x ] I have updated the documentation accordingly.
- [x ] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x ] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: keyboards/kalkulators/calc.c
Comment (waffle87) [inline keyboards/kalkulators/calc.c:1]: This file should be renamed to `kalkulators.c`.

File: keyboards/kalkulators/calc.c [hunk 1]
```diff
@@ -0,0 +1,454 @@
...
+#include <stdio.h>
+#include <stdlib.h>
+#include <math.h>
+#include <ctype.h>
+#include <string.h>
+#include "halconf.h"
+#include "tm1638.h"
+#include "usb_util.h"
+#include "wait.h"
+#include "calc.h"
```
Comment (waffle87) [inline keyboards/kalkulators/calc.c:27]:
```suggestion
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <ctype.h>
#include <string.h>
#include "halconf.h"
#include "tm1638.h"
#include "usb_util.h"
#include "wait.h"
#include "kalkulators.h"
```

File: keyboards/kalkulators/calc.c [hunk 2]
```diff
@@ -0,0 +1,454 @@
...
+        } else if (inputString[i] == '.') {
+            decimalFound = 1; // Decimal point is found
+        } else {
+            // Break the loop if a non-digit or non-decimal point character is encountered
+            break;
+        }
+    }
+}
+
+enum custom_keycodes { KC_cl = QK_USER, KC_div, KC_times, KC_minus, KC_seven, KC_eight, KC_nine, KC_four, KC_five, KC_six, KC_plus, KC_one, KC_two, KC_three, KC_zero, KC_dot, KC_equals };
```
Comment (waffle87) [inline keyboards/kalkulators/calc.c:66]:
```suggestion
enum custom_keycodes {
  KC_cl = QK_KB_0,
  KC_div,
  KC_times,
  KC_minus,
  KC_seven,
  KC_eight,
  KC_nine,
  KC_four,
  KC_five,
  KC_six,
  KC_plus,
  KC_one,
  KC_two,
  KC_three,
  KC_zero,
  KC_dot,
  KC_equals
};
```

File: keyboards/kalkulators/calc.c [hunk 3]
```diff
@@ -0,0 +1,454 @@
...
+    else if (input == 999) {
+        equals_operation();
+        if (!operation_state.waiting_for_operator) {
+            operation_state.enter_pressed = true; // Set the flag only if a valid calculation was performed
+        }
+    }
+}
+
+// listens for keypresses
+bool process_record_user(uint16_t keycode, keyrecord_t *record) {
```
Comment (waffle87) [inline keyboards/kalkulators/calc.c:392]:
```suggestion
bool process_record_kb(uint16_t keycode, keyrecord_t *record) {
```

File: keyboards/kalkulators/calc.c [hunk 4]
```diff
@@ -0,0 +1,454 @@
...
+                process_input(999);
+                operation_state.enter_pressed = true;
+
+                break;
+        }
+        return true;
+    }
+
+    return true;
+
```
Comment (waffle87) [inline keyboards/kalkulators/calc.c:453]:
```suggestion
    }
    return process_record_user(keycode, record);
```

File: keyboards/kalkulators/calc.h
Comment (waffle87) [inline keyboards/kalkulators/calc.h:1]: This file should be renamed to `kalkulators.h`.

File: keyboards/kalkulators/config.h [hunk 5]
```diff
@@ -0,0 +1,17 @@
+// Copyright 2023 georgsnarbuts (@georgsnarbuts)
+// SPDX-License-Identifier: GPL-2.0-or-later
+
+#pragma once
+
+/*
+ * Feature disable options
+ *  These options are also useful to firmware size reduction.
+ */
+
```
Comment (waffle87) [inline keyboards/kalkulators/config.h:10]:
```suggestion
```

File: keyboards/kalkulators/info.json [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+                {"matrix": [1, 3], "x": 3, "y": 1, "h": 2},
+                {"matrix": [3, 0], "x": 0, "y": 3},
+                {"matrix": [3, 1], "x": 1, "y": 3},
+                {"matrix": [3, 2], "x": 2, "y": 3},
+                {"matrix": [4, 0], "x": 0, "y": 4, "w": 2},
+                {"matrix": [4, 2], "x": 2, "y": 4},
+                {"matrix": [3, 3], "x": 3, "y": 3, "h": 2}
+            ]
+        }
+    }
```
Comment (waffle87) [inline keyboards/kalkulators/info.json:69]: How do these differ? They appear to be identical.

File: keyboards/kalkulators/keymaps/default/keymap.c
Comment (waffle87) [inline keyboards/kalkulators/keymaps/default/keymap.c:1]: Default keymaps are meant to be pristine, meaning much of this needs to be removed, like the combos and other custom code. If it is essential to device's functionality (eg. hardware code), it should be moved to the keyboard level in the `kalkulators.c` file.

File: keyboards/kalkulators/keymaps/via/keymap.c
Comment (waffle87) [inline keyboards/kalkulators/keymaps/via/keymap.c:1]: As above.

File: keyboards/kalkulators/readme.md [hunk 7]
```diff
@@ -0,0 +1,46 @@
+# kalkulators
+
+![kalkulators](https://i.imgur.com/hodQbV1.jpeg)
```
Comment (waffle87) [inline keyboards/kalkulators/readme.md:3]:
```suggestion
![kalkulators](https://i.imgur.com/hodQbV1h.jpeg)
```

File: keyboards/kalkulators/readme.md [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+An 8 digit 7-segment display calculator and numpad. 
+USB-C, STM32F072, TM1638 LED controller, 1xAA battery.
+
+**When editing layers through keymap or VIA, do not delete/edit the 1st layer, it is needed for the calculator functionality!**
+
+* Keyboard Maintainer: [georgsnarbuts](https://github.com/georgsnarbuts)
+
+Make example for this keyboard (after setting up your build environment):
+
+    qmk compile -kb kalkulators -km via
```
Comment (waffle87) [inline keyboards/kalkulators/readme.md:14]:
```suggestion
    make kalkulators:default
```

File: keyboards/kalkulators/rules.mk [hunk 9]
```diff
@@ -0,0 +1,3 @@
+# This file intentionally left blank
+SRC += tm1638.c calc.c
+NO_USB_STARTUP_CHECK = yes
```
Comment (waffle87) [inline keyboards/kalkulators/rules.mk:3]:
```suggestion
SRC += tm1638.c
```

File: keyboards/kalkulators/info.json [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+    "matrix_pins": {
+        "cols": ["A6", "A5", "A4", "A3"],
+        "rows": ["A7", "B0", "A2", "A1", "A0"]
+    },
+    "processor": "STM32F072",
+    "url": "",
+    "usb": {
+        "device_version": "1.0.0",
+        "pid": "0x4B4C",
+        "vid": "0x474E"
```
Comment (waffle87) [inline keyboards/kalkulators/info.json:24]:
```suggestion
        "vid": "0x474E",
        "no_startup_check": true
```

File: keyboards/kalkulators/tm1638.c [hunk 11]
```diff
@@ -0,0 +1,319 @@
...
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <stdint.h>
+#include <stdbool.h>
+#include "tm1638.h"
+#include "gpio.h"
+#include "progmem.h"
+#include "wait.h"
+#include QMK_KEYBOARD_H
```
Comment (waffle87) [inline keyboards/kalkulators/tm1638.c:23]:
```suggestion
#include "kalkulators.h"
```

**Review by waffle87 — CHANGES_REQUESTED** [review]

File: keyboards/kalkulators/calc.c (hunk 1 above)
Comment (georgsnarbuts) [reply keyboards/kalkulators/calc.c:27]: Is there something to replace "#include QMK_KEYBOARD_H" with? Removing it breaks the code.

File: keyboards/kalkulators/keymaps/default/keymap.c
Comment (georgsnarbuts) [reply keyboards/kalkulators/keymaps/default/keymap.c:1]: Combos don't seem to work outside of keymap.c . Are there ways to make it work or userspace should be used?

File: keyboards/kalkulators/calc.c (hunk 1 above)
Comment (waffle87) [reply keyboards/kalkulators/calc.c]: From my comment:
`#include "kalkulators.h"`
which is what `QMK_KEYBOARD_H` is, but the macro should not be used at the keyboard level.

File: keyboards/kalkulators/keymaps/default/keymap.c
Comment (waffle87) [reply keyboards/kalkulators/keymaps/default/keymap.c:1]: The latter, combos cannot be at the keyboard level as they typically require keymap specific configuration.

File: keyboards/kalkulators/chconf.h [hunk 12]
```diff
@@ -0,0 +1,21 @@
...
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+ 
+#pragma once
+
+#define CH_CFG_ST_FREQUENCY 10000
```
Comment (waffle87) [inline keyboards/kalkulators/chconf.h:19]: This is the default value, so this file can be removed.

File: keyboards/kalkulators/keymaps/default/keymap.c [hunk 13]
```diff
@@ -0,0 +1,274 @@
+// Copyright 2023 QMK
+// SPDX-License-Identifier: GPL-2.0-or-later
+
+#include QMK_KEYBOARD_H
+#include <stdio.h>
+#include "halconf.h"
+#include "tm1638.h"
+#include "wait.h"
+#include "usb_util.h"
+#include "kalkulators.h"
```
Comment (waffle87) [inline keyboards/kalkulators/keymaps/default/keymap.c:10]:
```suggestion
```

File: keyboards/kalkulators/info.json [hunk 14]
```diff
@@ -0,0 +1,50 @@
...
+    "processor": "STM32F072",
+    "url": "",
+    "usb": {
+        "device_version": "1.0.0",
+        "pid": "0x4B4C",
+        "vid": "0x474E",
+        "no_startup_check": true
+    },
+    "layouts": {
+        "LAYOUT_calc": {
```
Comment (waffle87) [inline keyboards/kalkulators/info.json:28]:
```suggestion
        "LAYOUT": {
```

File: keyboards/kalkulators/keymaps/default/keymap.c [hunk 15]
```diff
@@ -0,0 +1,274 @@
...
+    [0] = LAYOUT_calc(
+        KC_DELETE,  KC_PSLS, KC_PAST, KC_PMNS,
+        KC_P7,   KC_P8,   KC_P9,
+        KC_P4,   KC_P5,   KC_P6,   KC_PPLS,
+        KC_P1,   KC_P2,   KC_P3,
+        KC_P0,   KC_PDOT,          KC_PENT
+    ),
+
+    [1] = LAYOUT_calc(
+        QK_KB_0,   KC_div, KC_times, KC_minus,
```
Comment (waffle87) [inline keyboards/kalkulators/keymaps/default/keymap.c:116]:
```suggestion
    [0] = LAYOUT(
        KC_DELETE,  KC_PSLS, KC_PAST, KC_PMNS,
        KC_P7,   KC_P8,   KC_P9,
        KC_P4,   KC_P5,   KC_P6,   KC_PPLS,
        KC_P1,   KC_P2,   KC_P3,
        KC_P0,   KC_PDOT,          KC_PENT
    ),

    [1] = LAYOUT(
        KC_cl,   KC_div, KC_times, KC_minus,
```

File: keyboards/kalkulators/keymaps/default/keymap.c [hunk 16]
```diff
@@ -0,0 +1,274 @@
...
+        KC_seven,   KC_eight,   KC_nine,
+        KC_four,   KC_five,   KC_six,   KC_plus,
+        KC_one,   KC_two,   KC_three,
+        KC_zero,   KC_dot,          KC_equals
+    )
+};
+
+void housekeeping_task_user(void) {
```
Comment (waffle87) [inline keyboards/kalkulators/keymaps/default/keymap.c:124]: Should also probably be moved to the keyboard level. Pretty much anything controlling hardware that doesn't depend on keymap-specific configuration should be at the keyboard level.

File: keyboards/kalkulators/keymaps/default/keymap.c [hunk 17]
```diff
@@ -0,0 +1,274 @@
...
+  KC_six,   
+  KC_plus,
+  KC_one,   
+  KC_two,   
+  KC_three,
+  KC_zero,   
+  KC_dot,          
+  KC_equals
+};
+
```
Comment (waffle87) [inline keyboards/kalkulators/keymaps/default/keymap.c:91]:
```suggestion
```

File: keyboards/kalkulators/kalkulators.c [hunk 18]
```diff
@@ -0,0 +1,471 @@
...
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include QMK_KEYBOARD_H
+#include <stdio.h>
+#include <stdlib.h>
+#include <math.h>
+#include <ctype.h>
+#include <string.h>
+#include "kalkulators.h"
```
Comment (waffle87) [inline keyboards/kalkulators/kalkulators.c:23]:
```suggestion
```

File: keyboards/kalkulators/keymaps/via/keymap.c [hunk 19]
```diff
@@ -0,0 +1,278 @@
+// Copyright 2023 QMK
+// SPDX-License-Identifier: GPL-2.0-or-later
+
+#include QMK_KEYBOARD_H
+#include <stdio.h>
+#include "halconf.h"
+#include "tm1638.h"
+#include "wait.h"
+#include "usb_util.h"
+#include "kalkulators.h"
```
Comment (waffle87) [inline keyboards/kalkulators/keymaps/via/keymap.c:10]:
```suggestion
```

File: keyboards/kalkulators/keymaps/via/keymap.c [hunk 20]
```diff
@@ -0,0 +1,278 @@
...
+
+int brightness_level = 7;
+
+bool show_usb_connection = true;
+bool show_calc_connection = true;
+bool pc_connection;
+int layer;
+
+void keyboard_post_init_user(void) {
```
Comment (waffle87) [inline keyboards/kalkulators/keymaps/via/keymap.c:19]: As above, w.r.t keyboard level.

File: keyboards/kalkulators/keymaps/via/keymap.c [hunk 21]
```diff
@@ -0,0 +1,278 @@
...
+  KC_plus,
+  KC_one,
+  KC_two,
+  KC_three,
+  KC_zero,
+  KC_dot,
+  KC_equals
+};
+
+
```
Comment (waffle87) [inline keyboards/kalkulators/keymaps/via/keymap.c:93]:
```suggestion
```

File: keyboards/kalkulators/keymaps/via/keymap.c [hunk 22]
```diff
@@ -0,0 +1,278 @@
...
+    [0] = LAYOUT_calc(
+        KC_DELETE,  KC_PSLS, KC_PAST, KC_PMNS,
+        KC_P7,   KC_P8,   KC_P9,
+        KC_P4,   KC_P5,   KC_P6,   KC_PPLS,
+        KC_P1,   KC_P2,   KC_P3,
+        KC_P0,   KC_PDOT,          KC_PENT
+    ),
+
+    [1] = LAYOUT_calc(
+        QK_KB_0,   KC_div, KC_times, KC_minus,
```
Comment (waffle87) [inline keyboards/kalkulators/keymaps/via/keymap.c:117]:
```suggestion
    [0] = LAYOUT(
        KC_DELETE,  KC_PSLS, KC_PAST, KC_PMNS,
        KC_P7,   KC_P8,   KC_P9,
        KC_P4,   KC_P5,   KC_P6,   KC_PPLS,
        KC_P1,   KC_P2,   KC_P3,
        KC_P0,   KC_PDOT,          KC_PENT
    ),

    [1] = LAYOUT(
        KC_cl,   KC_div, KC_times, KC_minus,
```

File: keyboards/kalkulators/keymaps/via/keymap.c [hunk 23]
```diff
@@ -0,0 +1,278 @@
...
+        KC_seven,   KC_eight,   KC_nine,
+        KC_four,   KC_five,   KC_six,   KC_plus,
+        KC_one,   KC_two,   KC_three,
+        KC_zero,   KC_dot,          KC_equals
+    )
+};
+
+
+
+void housekeeping_task_user(void) {
```
Comment (waffle87) [inline keyboards/kalkulators/keymaps/via/keymap.c:127]: Duplicate of above `housekeeping_task_user`?

**Review by waffle87 — CHANGES_REQUESTED** [review]

File: keyboards/kalkulators/info.json (hunk 14 above)
Comment (drashna) [reply keyboards/kalkulators/info.json:28]: `LAYOUT_numpad_5x4` would be more appropriate

**Pull request closed without merging**